src/crimson/os/seastore/transaction_manager.cc

```diff
@@ -682,12 +682,21 @@ TransactionManager::get_extents_if_live(
             auto &pin_seg_paddr = pin_paddr.as_seg_paddr();
             auto pin_paddr_seg_id = pin_seg_paddr.get_segment_id();
             auto pin_len = pin->get_length();
-            if (pin_paddr_seg_id != paddr_seg_id) {
+            // pin may be out of the range paddr~len, consider the following scene:
+            // 1. Trans.A writes the final record of Segment S, in which it overwrite
+            //    another extent E in the same segment S;
+            // 2. Before Trans.A "complete_commit", Trans.B tries to rewrite new
+            //    records and roll the segments, which closes Segment S;
+            // 3. Before Trans.A "complete_commit", a new cleaner Transaction C tries
+            //    to clean the segment;
+            //
+            // In this scenario, C might see a part of extent E's laddr space mapped
+            // to another location within the same segment S.
+            if (bool pin_valid = pin_seg_paddr >= paddr &&
+                  pin_seg_paddr.add_offset(pin_len) <= paddr.add_offset(len);
```

Comment:
IIUC this is a valid case even if pin-paddr is out of the expected range but exists in the same segment.

So this check (`pin_valid`) should be removed, right?

Reply:
Yes, that's right. But this might be contraditory with https://tracker.ceph.com/issues/66941, if we decide to make reclaiming recognize committed segments, pin-paddr can't be out of the expected range. Am I right? @zhscn

Reply:
Seems to me the fix to https://tracker.ceph.com/issues/66941 can be done independently. After that, we should be able to revert this PR and recover the original `ceph_assert` that the extent is impossible to be outside the expected range and also in the same segment.

Reply:
Yeah, that's correct:-)